oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/version/VersionableState.java

```diff
@@ -285,8 +287,32 @@ private void restoreFrozen(@NotNull NodeBuilder frozen,
                                @NotNull NodeBuilder dest,
                                @NotNull VersionSelector selector)
             throws RepositoryException, CommitFailedException {
-        // 15.7.2 Restoring Type and Identifier
-        restoreFrozenTypeAndUUID(frozen, dest);
+        //OAK-9459: if the NodeState is not empty, retrieve OPVs before restoring types to avoid constraint violations
+        boolean frozenTypeRestored = false;
+        if (!dest.hasProperty(JCR_PRIMARYTYPE)) {
+            // empty NodeState
+            // 15.7.2 Restoring Type and Identifier
+            restoreFrozenTypeAndUUID(frozen, dest);
+            frozenTypeRestored = true;
+        }
+        HashMap<PropertyState, Integer> opvs = new HashMap<>();
+        for (PropertyState p : dest.getProperties()) {
+            String propName = p.getName();
+            if (BASIC_PROPERTIES.contains(propName)) {
+                continue;
```

Comment:
Can we change the logic not to use "continue"?

Reply:
Certainly, I did not touch the logic here. What is wrong with continue?

Reply:
I don't care, so I'll remove the continue and and I think we should also turn the if-else if sequence into a switch.